File: src/lib/steg/crypto.ts
/**
 * Optional confidentiality layer: PBKDF2-HMAC-SHA256 + AES-256-GCM, on WebCrypto.
 *
 * **Steganography hides data, it does not protect it.** Anyone who can run the decoder
 * reads whatever was embedded. When the user supplies a passphrase the payload is sealed
 * into an authenticated envelope first, so the carrier is worthless on its own.
 *
 * Byte-compatible with `crates/steg-core/src/crypto.rs`; see
 * `docs/STEGO_FORMAT.md` §2 for the layout.
 */

import { ENVELOPE_HEADER_LEN, ENVELOPE_MAGIC, indexOfBytes } from './core.ts';
import { StegError, StegErrorCode } from './types.ts';

/** PBKDF2 rounds. Matches `KDF_ITERATIONS` in the Rust core. */
export const KDF_ITERATIONS = 210_000;

const SALT_LEN = 32;
const IV_LEN = 12;

const encoder = new TextEncoder();

/**
 * Additional authenticated data. Binding the envelope to a format version stops a valid
 * envelope from being transplanted into a future container version.
 */
const AAD = encoder.encode('masc-steg-v1');

/**
 * TypeScript 5.7 made typed arrays generic over their backing buffer, and the DOM lib
 * types `BufferSource` as `ArrayBufferView<ArrayBuffer>`. Every array we hand WebCrypto
 * is created here or in our own engine, so it is always `ArrayBuffer`-backed; this
 * narrows the type without changing behaviour.
 */
function src(bytes: Uint8Array): Uint8Array<ArrayBuffer> {
	return bytes as Uint8Array<ArrayBuffer>;
}

/** True when `crypto.subtle` is usable — requires a secure context (HTTPS or localhost). */
export function cryptoAvailable(): boolean {
	return typeof crypto !== 'undefined' && typeof crypto.subtle !== 'undefined';
}

function requireCrypto(): SubtleCrypto {
	if (!cryptoAvailable()) {
		throw new StegError(
			StegErrorCode.Crypto,
			'WebCrypto is unavailable. The passphrase layer needs a secure context (https:// or localhost).'
		);
	}
	return crypto.subtle;
}

/**
 * Seal `plaintext` into a `STGECRY1` envelope.
 *
 * Salt and IV come from the platform CSPRNG, never from `Math.random()`.
 */
export async function seal(
	plaintext: Uint8Array,
	passphrase: string,
	iterations: number = KDF_ITERATIONS
): Promise<Uint8Array> {
	const subtle = requireCrypto();
	if (!passphrase) {
		throw new StegError(StegErrorCode.Crypto, 'a passphrase is required to encrypt the payload');
	}

	const salt = crypto.getRandomValues(new Uint8Array(SALT_LEN));
	const iv = crypto.getRandomValues(new Uint8Array(IV_LEN));

	const key = await deriveKey(passphrase, salt, iterations);
	const ciphertext = new Uint8Array(
		await subtle.encrypt(
			{ name: 'AES-GCM', iv: src(iv), additionalData: AAD, tagLength: 128 },
			key,
			src(plaintext)
		)
	);

	const out = new Uint8Array(ENVELOPE_HEADER_LEN + ciphertext.length);
	out.set(encoder.encode(ENVELOPE_MAGIC), 0);
	new DataView(out.buffer).setUint16(8, iterations, true);
	out[10] = SALT_LEN;
	out.set(salt, 11);
	out[43] = IV_LEN;
	out.set(iv, 44);
	new DataView(out.buffer).setUint32(56, ciphertext.length, true);
	out.set(ciphertext, ENVELOPE_HEADER_LEN);
	return out;
}

/**
 * Open a `STGECRY1` envelope.
 *
 * A wrong passphrase and a tampered payload both throw `BAD_PASSPHRASE`. GCM genuinely
 * cannot tell them apart, and reporting which one it was would hand an attacker an
 * oracle.
 */
export async function unseal(envelope: Uint8Array, passphrase: string): Promise<Uint8Array> {
	const subtle = requireCrypto();
	if (!isEnvelope(envelope)) {
		throw new StegError(StegErrorCode.Malformed, 'not a STGECRY1 envelope');
	}

	const view = new DataView(envelope.buffer, envelope.byteOffset, envelope.byteLength);
	const iterations = view.getUint16(8, true);
	if (iterations === 0) {
		throw new StegError(StegErrorCode.Malformed, 'envelope declares 0 PBKDF2 iterations');
	}

	const saltLen = envelope[10];
	if (saltLen === 0 || saltLen > 64) {
		throw new StegError(StegErrorCode.Malformed, `bad salt length ${saltLen}`);
	}
	const ivAt = 11 + saltLen;
	if (ivAt + 1 > envelope.length) {
		throw new StegError(StegErrorCode.Truncated, 'Unexpected end of input');
	}
	const ivLen = envelope[ivAt];
	if (ivLen !== IV_LEN) {
		throw new StegError(StegErrorCode.Malformed, `bad IV length ${ivLen}`);
	}
	const lenAt = ivAt + 1 + ivLen;
	if (lenAt + 4 > envelope.length) {
		throw new StegError(StegErrorCode.Truncated, 'Unexpected end of input');
	}
	const ctLen = view.getUint32(lenAt, true);
	if (lenAt + 4 + ctLen !== envelope.length) {
		throw new StegError(
			StegErrorCode.Malformed,
			'ciphertext length does not match envelope size'
		);
	}

	const salt = envelope.subarray(11, ivAt);
	const iv = envelope.subarray(ivAt + 1, lenAt);
	const ciphertext = envelope.subarray(lenAt + 4);

	try {
		const key = await deriveKey(passphrase, salt, iterations);
		const plaintext = await subtle.decrypt(
			{ name: 'AES-GCM', iv: src(iv), additionalData: AAD, tagLength: 128 },
			key,
			src(ciphertext)
		);
		return new Uint8Array(plaintext);
	} catch (err) {
		if (err instanceof StegError) throw err;
		throw new StegError(
			StegErrorCode.BadPassphrase,
			'wrong passphrase, or the payload was re-compressed'
		);
	}
}

/** Cheap check so the UI can label a payload without attempting decryption. */
export function isEnvelope(bytes: Uint8Array): boolean {
	if (bytes.length < ENVELOPE_HEADER_LEN) return false;
	// Compare the magic without allocating a decoder round trip.
	return indexOfBytes(bytes, encoder.encode(ENVELOPE_MAGIC), 0) === 0;
}

async function deriveKey(
	passphrase: string,
	salt: Uint8Array,
	iterations: number
): Promise<CryptoKey> {
	const subtle = requireCrypto();
	// Import the passphrase first so PBKDF2 does not have to re-encode it.
	const material = await subtle.importKey('raw', encoder.encode(passphrase), 'PBKDF2', false, [
		'deriveKey'
	]);
	return subtle.deriveKey(
		{ name: 'PBKDF2', salt: src(salt), iterations, hash: 'SHA-256' },
		material,
		{ name: 'AES-GCM', length: 256 },
		false,
		['encrypt', 'decrypt']
	);
}

File: crates/steg-core/src/crypto.rs
//! Optional confidentiality layer: PBKDF2-HMAC-SHA256 + AES-256-GCM.
//!
//! **Steganography hides data; it does not protect it.** Anyone who can run the decoder
//! can read whatever was embedded. When the user supplies a passphrase, the payload is
//! sealed into an authenticated envelope first, so the carrier is useless on its own.
//!
//! Layout is normative in `docs/STEGO_FORMAT.md` §2. The TypeScript mirror in
//! `src/lib/steg/crypto.ts` must produce byte-identical envelopes.

use crate::format::ENVELOPE_MAGIC;
use crate::{Error, Result};
use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use hmac::Hmac;
use sha2::Sha256;

/// PBKDF2 rounds used when the caller does not override them.
pub const KDF_ITERATIONS: u16 = 210_000;

const SALT_LEN: usize = 32;
const IV_LEN: usize = 12;
const FIXED_HEADER_LEN: usize = 60;
const MAX_ENVELOPE_LEN: usize = 1 << 30;

/// Additional authenticated data. Binding the envelope to a format version stops a
/// valid envelope from being transplanted into a future container version.
const AAD: &[u8] = b"masc-steg-v1";

/// Source of cryptographically secure random bytes.
///
/// Abstracted so the pure logic can be unit-tested on a machine with no OS CSPRNG, and
/// so the WASM layer can use the browser's `crypto.getRandomValues`.
pub trait Rng {
    fn fill(&mut self, buf: &mut [u8]);
}

/// An RNG that always returns the same bytes. **Test use only.**
pub struct FixedRng(pub u8);
impl Rng for FixedRng {
    fn fill(&mut self, buf: &mut [u8]) {
        buf.fill(self.0);
    }
}

/// Seal `plaintext` into an authenticated envelope using `passphrase`.
///
/// `iterations` is written into the header so the round count can be raised later
/// without breaking old containers.
pub fn seal(plaintext: &[u8], passphrase: &[u8], iterations: u16, rng: &mut dyn Rng) -> Result<Vec<u8>> {
    if iterations == 0 {
        return Err(Error::Crypto("refusing to run PBKDF2 with 0 iterations".into()));
    }

    let mut salt = [0u8; SALT_LEN];
    let mut iv = [0u8; IV_LEN];
    rng.fill(&mut salt);
    rng.fill(&mut iv);

    let key = derive_key(passphrase, &salt, iterations);
    let cipher = Aes256Gcm::new_from_slice(&key)
        .map_err(|e| Error::Crypto(format!("key init failed: {e}")))?;

    // The 16-byte GCM tag is appended by `encrypt`.
    let ct = cipher
        .encrypt(Nonce::from_slice(&iv), Payload { msg: plaintext, aad: AAD })
        .map_err(|e| Error::Crypto(format!("encrypt failed: {e}")))?;

    let mut out = Vec::with_capacity(FIXED_HEADER_LEN + ct.len());
    out.extend_from_slice(ENVELOPE_MAGIC);
    out.extend_from_slice(&iterations.to_le_bytes());
    out.push(SALT_LEN as u8);
    out.extend_from_slice(&salt);
    out.push(IV_LEN as u8);
    out.extend_from_slice(&iv);
    out.extend_from_slice(&(ct.len() as u32).to_le_bytes());
    out.extend_from_slice(&ct);
    debug_assert_eq!(out.len(), FIXED_HEADER_LEN + ct.len());
    Ok(out)
}

/// Open an envelope produced by [`seal`].
///
/// A wrong passphrase and a tampered ciphertext both surface as
/// [`Error::BadPassphrase`] on purpose — GCM cannot tell them apart, and telling the
/// user which one it was would be an oracle.
pub fn unseal(envelope: &[u8], passphrase: &[u8]) -> Result<Vec<u8>> {
    if envelope.len() < FIXED_HEADER_LEN || &envelope[..8] != ENVELOPE_MAGIC {
        return Err(Error::Malformed("not a STGECRY1 envelope".into()));
    }
    if envelope.len() > MAX_ENVELOPE_LEN {
        return Err(Error::Malformed("envelope is implausibly large".into()));
    }

    let iterations = u16::from_le_bytes([envelope[8], envelope[9]]);
    if iterations == 0 {
        return Err(Error::Malformed("envelope declares 0 PBKDF2 iterations".into()));
    }

    let salt_len = envelope[10] as usize;
    if salt_len == 0 || salt_len > 64 {
        return Err(Error::Malformed(format!("bad salt length {salt_len}")));
    }
    let iv_at = 11 + salt_len;
    if iv_at + 1 > envelope.len() {
        return Err(Error::Truncated);
    }
    let iv_len = envelope[iv_at] as usize;
    if iv_len != IV_LEN {
        return Err(Error::Malformed(format!("bad IV length {iv_len}")));
    }
    let len_at = iv_at + 1 + iv_len;
    if len_at + 4 > envelope.len() {
        return Err(Error::Truncated);
    }
    let ct_len = u32::from_le_bytes([
        envelope[len_at],
        envelope[len_at + 1],
        envelope[len_at + 2],
        envelope[len_at + 3],
    ]) as usize;
    let ct_start = len_at + 4;
    if ct_start + ct_len != envelope.len() {
        return Err(Error::Malformed("ciphertext length does not match envelope size".into()));
    }

    let salt = &envelope[11..iv_at];
    let iv = &envelope[iv_at + 1..len_at];
    let key = derive_key(passphrase, salt, iterations);
    let cipher =
        Aes256Gcm::new_from_slice(&key).map_err(|e| Error::Crypto(format!("key init: {e}")))?;

    cipher
        .decrypt(
            Nonce::from_slice(iv),
            Payload { msg: &envelope[ct_start..], aad: AAD },
        )
        .map_err(|_| Error::BadPassphrase)
}

/// True when `bytes` looks like a `STGECRY1` envelope.
pub fn is_envelope(bytes: &[u8]) -> bool {
    bytes.len() >= FIXED_HEADER_LEN && &bytes[..8] == ENVELOPE_MAGIC
}

fn derive_key(passphrase: &[u8], salt: &[u8], iterations: u16) -> [u8; 32] {
    let mut key = [0u8; 32];
    pbkdf2::pbkdf2::<Hmac<Sha256>>(passphrase, salt, iterations as u32, &mut key);
    key
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip() {
        let mut rng = FixedRng(0xA7);
        let sealed = seal(b"the eagle lands at dawn", b"correct horse", 1_000, &mut rng).unwrap();
        assert!(is_envelope(&sealed));
        let opened = unseal(&sealed, b"correct horse").unwrap();
        assert_eq!(opened, b"the eagle lands at dawn");
    }

    #[test]
    fn wrong_passphrase_is_rejected() {
        let mut rng = FixedRng(0x11);
        let sealed = seal(b"secret", b"right", 1_000, &mut rng).unwrap();
        assert_eq!(unseal(&sealed, b"wrong").unwrap_err(), Error::BadPassphrase);
    }

    #[test]
    fn tampered_ciphertext_is_rejected() {
        let mut rng = FixedRng(0x22);
        let mut sealed = seal(b"secret", b"right", 1_000, &mut rng).unwrap();
        let last = sealed.len() - 1;
        sealed[last] ^= 0x01;
        assert_eq!(unseal(&sealed, b"right").unwrap_err(), Error::BadPassphrase);
    }

    #[test]
    fn deterministic_kdf_matches_known_vector() {
        // RFC 6070-style check adapted to SHA-256, vector from the PBKDF2-HMAC-SHA256
        // test suite: password "password", salt "salt", 1 iteration.
        let key = derive_key(b"password", b"salt", 1);
        assert_eq!(
            key.iter().map(|b| format!("{b:02x}")).collect::<String>(),
            "120fb6cffcf8b32c43e7225256c4f837a86548c92ccc35480805987cb70be17b"
        );
    }
}
